freqtrade/exchange/hyperliquid.py

```diff
@@ -15,21 +19,145 @@ class Hyperliquid(Exchange):
     Contains adjustments needed for Freqtrade to work with this exchange.
     """
 
-    _ft_has: FtHas = {
-        # Only the most recent 5000 candles are available according to the
-        # exchange's API documentation.
+    _ft_has: Dict = {
         "ohlcv_has_history": False,
         "ohlcv_candle_limit": 5000,
-        "trades_has_history": False,  # Trades endpoint doesn't seem available.
+        "orderbook_max_entries": 20,
+        "l2_limit_range": [20],
+        "trades_has_history": False,
+        "tickers_have_bid_ask": False,
+        "stoploss_on_exchange": True,
         "exchange_has_overrides": {"fetchTrades": False},
+        "stoploss_order_types": {"limit": "limit"},
+        "funding_fee_timeframe": "1h",
+        "create_order_has_all_data": False
     }
 
+    _supported_trading_mode_margin_pairs: List[Tuple[TradingMode, MarginMode]] = [
+        (TradingMode.FUTURES, MarginMode.ISOLATED)
+    ]
+
     @property
-    def _ccxt_config(self) -> dict:
-        # Parameters to add directly to ccxt sync/async initialization.
-        # ccxt defaults to swap mode.
+    def _ccxt_config(self) -> Dict:
+        # ccxt Hyperliquid defaults to swap
         config = {}
         if self.trading_mode == TradingMode.SPOT:
             config.update({"options": {"defaultType": "spot"}})
         config.update(super()._ccxt_config)
         return config
+
+    def get_max_leverage(self, pair: str, stake_amount: Optional[float]) -> float:
+        # There are no leverage tiers
+        if self.trading_mode == TradingMode.FUTURES:
+            return self.markets[pair]['limits']['leverage']['max']
+        else:
+            return 1.0
+
+    def ohlcv_candle_limit(
+        self, timeframe: str, candle_type: CandleType, since_ms: Optional[int] = None
+    ) -> int:
+        # Funding rate candles have a different limit
+        if candle_type in CandleType.FUNDING_RATE:
+            return 500
+
+        return super().ohlcv_candle_limit(timeframe, candle_type, since_ms)
+
+    def _lev_prep(self, pair: str, leverage: float, side: BuySell, accept_fail: bool = False):
+        if self.trading_mode != TradingMode.SPOT:
+            # Hyperliquid expects leverage to be an int
+            leverage = int(leverage)
+            # Hyperliquid needs the parameter leverage.
+            # Don't use set_leverage(), as this sets margin back to cross
+            self.set_margin_mode(pair, self.margin_mode, params={"leverage": leverage})
+
+    def dry_run_liquidation_price(
```

Comment:
This will for sure need a dedicated test - ideally a parametrized one testing multiple scenarios (long,short, ...).

ideally, the exchange docs have a calculation example you can "copy" from (one that contains the necessary values and expected results) - which should be one of the tests.

Reply:
ok! I followed the docs, so I think that's good. Tested it against ~170 unique ccxt `fetch_positions()` outputs on real orders, I'll make a test for that.

Reply:
some info about why I chose those positions:
- different prices (BTC is high, DOGE is low, SOL & ETH in between)
- different sides
- different leverages (1-5)

max 1 position per symbol/side/leverage combination. Could remove some if you think this is overkill.